DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,308 @@
+
+Avoid stutter in struct and interface names by treating the package name as a namespace and considering how the types are consumed. For example, `drink.DrinkDecorator` should be `drink.Decorator`. You may find you have a type named after the package, soemtimes it's appropriate.
+
+If you find yourself using an appropriate package name that's commonly used as a variable, consider pluralising the package name.
+
+Avoid [relentlessly] `types`, `interfaces`, `common`, `util` or `base` packages. These names offer no indication on what the package provides beyond 'a bundle of stuff'. With time, these types of package turn into dumping grounds containing unrelated logic that can be the root cause for cyclic dependencies.  If a utility style package is required, make it specific. For example a `/utils/cmdutil` package may provide command utility functions.
+
+##### Function & method
+
+Functions and methods (functions with a receiver parameter) should adequetly describe their behavior. Generally, they should follow a verb-noun form. The types of parameters should help inform the function name.
+
+##### Variable
+
+Variable names should be concise and descriptive. Prefer single word names. The further away from the site of declaration a variable is used, the more descriptive it needs to be. For example, a variable named `n` used 30 lines after declaration makes it unnecessarily difficult to reason what it represents at the site of use.
+
+### <a name="dependencies"></a> Dependencies
+
+Functions should explicitly declare their dependencies as parameters. In doing so, they honor the [dependency inversion principle](https://www.digitalocean.com/community/conceptual_articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#dependency-inversion-principle) and provide the necessary flexibility to write loosely coupled programs.
+
+```go
+func FastenPlanks(tools Toolbox, surface Surface, planks Planks) error {
+    for ; planks.Count() > 0;  {
+        if err := tools.Hammer(surface, planks.Next()); err != nil {
+            return err
+        }
+    }
+}
+```
+
+Avoid constructing dependencies in-line. If you do construct a dependency in-line because it compliments the most common usage, provide a way to override it such as with a setter.
+
+##### "Accept interfaces, return structs"
+
+The Go community talks of the "accept interfaces, return structs" addage. The meaning of this idiom can be hard to pin down, herein lies some clarity.
+
+"Accept interfaces" speaks to the Dependency Inversion Principle (D in 'SOLID'). When a function declares dependent behavior as an interface parameter it gives control of _how_ that behavior is satisfied to the consuming logic. This is the foundation of flexible code.
```

Comment:
Repeating myself here a bit but: using interfaces in go is key for the Dependency Inversion Principle, but not enough. What that interface represents and the signature of its methods play a key role as well.

Maybe this is something we can elaborate more on the interfaces section: using an interface is not enough to achieve loosely coupling.

Reply:
I'm not sure I agree. Dependency inversion can be achieved with an interface parameter. I don't think we should complicate it beyond that. Whether your _design_ is providing increased flexibility is related but different. Your interface designs speaks to the _degree_ of flexibility but the degree of flexibility goes beyond dependency inversion.

Reply:
My bad, I think I didn't explain myself correctly. I might do better with an example:

```go
# mysql/auth.go

type Credentials struct {
   User, Password string
}

func (a *Authenticator) Authenticate(ctx context.Context, creds Credentials) error {...}


# user/registry.go

type DBConnector interface {
    Authenticate(ctx context.Context, creds Credentials) error
}

type Registry struct {
   dBConnector DBConnector
}

func NewRegistry(dBConnector DBConnector) *Registry {...}

func (r *Registry) Start(ctx context.Context) error {
   ...
   err := r.Authenticate(ctx, mysql.Credentials{User: user, Password: password})
   ...
}
```

Here, the registry is taking its dependency as an interface, however, it doesn't achieve dependency inversion because it still depends on a type from the concrete implementation. This is what I meant that just using an interface is not enough. You need to design the interface appropriately, for example:

```go
# mysql/auth.go
func (a *Authenticator) Authenticate(ctx context.Context, user, password string) error {...}


# user/registry.go
type DBConnector interface {
    Authenticate(ctx context.Context, user, password string) error
}

func (r *Registry) Start(ctx context.Context) error {
   ...
   err := r.Authenticate(ctx, user, password)
   ...
}
```

Reply:
I think that example alludes to different degrees of coupling which, while related, isn't strictly required to satisfy the dependency inversion principle. At the most basic level, dependency inversion can still be achieved for the purpose of testability by accepting an interface parameter. That interface may provide behavior that accepts types associated with a specific package. Yes, its advantageous to only use standard library types on interfaces and when you model complete abstractions a set of shared types but that again starts to delve into software design.

Given the contention I think I'll alter the wording by replacing any reference to DIP with 'loose coupling' or similar. The DIP commentary isn't necessary to convey the point.

Reply:
Sounds good, I'm ok with removing the DIP reference

Reply:
Addressed.